File: 20-ValidParentheses/main.go
package main

import "fmt"

func isValid(s string) bool {
	pairs := map[rune]rune{
		')': '(',
		']': '[',
		'}': '{',
	}

	stack := []rune{}

	for _, char := range s {
		if open, isClosing := pairs[char]; isClosing {
			if len(stack) == 0 || stack[len(stack)-1] != open {
				return false
			}

			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, char)
		}
	}

	return len(stack) == 0
}

func main() {
	fmt.Println(isValid("()"))
	fmt.Println(isValid("()[]{}"))
	fmt.Println(isValid("(]"))
	fmt.Println(isValid("([])"))
}

File: 20-ValidParentheses/Main.java
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

class ValidParentheses {
  public boolean isValid(String s) {
    Map<Character, Character> pairs = new HashMap<>();
    pairs.put(')', '(');
    pairs.put(']', '[');
    pairs.put('}', '{');

    Stack<Character> stack = new Stack<>();

    for (char c: s.toCharArray()) {
      if (pairs.containsKey(c)) {
        if (stack.isEmpty() || stack.peek() != pairs.get(c)) {
          return false;
        }

        stack.pop();
      } else {
        stack.add(c);
      }
    }

    return stack.isEmpty();
  }

  public static void main(String[] args) {
    ValidParentheses sol = new ValidParentheses();
    System.out.println(sol.isValid("()"));
    System.out.println(sol.isValid("()[]{}"));
    System.out.println(sol.isValid("(]"));
    System.out.println(sol.isValid("([])"));
  }
}